presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java

```diff
@@ -424,18 +561,78 @@ private boolean isCandidate(Map<ColumnHandle, NullableValue> bindings)
+            LookupVariableResolver inputs = new LookupVariableResolver(assignments, bindings, variable -> new Symbol(variable.getName()).toSymbolReference());
 
             // Skip pruning if evaluation fails in a recoverable way. Failing here can cause
             // spurious query failures for partitions that would otherwise be filtered out.
             Object optimized = TryFunction.evaluate(() -> evaluator.optimize(inputs), true);
 
             // If any conjuncts evaluate to FALSE or null, then the whole predicate will never be true and so the partition should be pruned
-            if (Boolean.FALSE.equals(optimized) || optimized == null || optimized instanceof NullLiteral) {
-                return false;
+            return !Boolean.FALSE.equals(optimized) && optimized != null && !(optimized instanceof NullLiteral);
+        }
+    }
+
+    private static class LayoutConstraintEvaluatorForRowExpression
+    {
+        private final Map<VariableReferenceExpression, ColumnHandle> assignments;
+        private final RowExpressionInterpreter evaluator;
+        private final Set<ColumnHandle> arguments;
+
+        public LayoutConstraintEvaluatorForRowExpression(Metadata metadata, Session session, Map<VariableReferenceExpression, ColumnHandle> assignments, RowExpression expression)
+        {
+            this.assignments = assignments;
+
+            evaluator = new RowExpressionInterpreter(expression, metadata, session.toConnectorSession(), OPTIMIZED);
+            arguments = VariablesExtractor.extractUnique(expression).stream()
+                    .map(assignments::get)
+                    .collect(toImmutableSet());
+        }
+
+        private boolean isCandidate(Map<ColumnHandle, NullableValue> bindings)
+        {
+            if (intersection(bindings.keySet(), arguments).isEmpty()) {
+                return true;
+            }
+            LookupVariableResolver inputs = new LookupVariableResolver(assignments, bindings, variable -> variable);
+
+            // Skip pruning if evaluation fails in a recoverable way. Failing here can cause
+            // spurious query failures for partitions that would otherwise be filtered out.
+            Object optimized = TryFunction.evaluate(() -> evaluator.optimize(inputs), true);
+
+            // If any conjuncts evaluate to FALSE or null, then the whole predicate will never be true and so the partition should be pruned
+            return !Boolean.FALSE.equals(optimized) && optimized != null && (!(optimized instanceof ConstantExpression) || !((ConstantExpression) optimized).isNull());
+        }
+    }
+
+    private static class LookupVariableResolver
+            implements VariableResolver
+    {
+        private final Map<VariableReferenceExpression, ColumnHandle> assignments;
+        private final Map<ColumnHandle, NullableValue> bindings;
+        private final Function<VariableReferenceExpression, Object> missingBindingSupplier;
```

Comment:
Add a comment why the value type is `Object`, and maybe a TODO stating change to `RowExpression` once all migration is done? 

Is it possible to make `LookupVariableResolver` a generic class?